### DISCUSSION TITLE
best way to add surname and lastname from 0auth

### DISCUSSION THREAD

**Discussion by Laestry** [pocketbase/pocketbase#6118] [Q&A] [answered]
What would be the best way to get the profile from google sign in and apple sign in?
I'm seeing the  "Optional user create fields map" for pocketbase 0.23.5
but it works only with google and I can't seperate sur/lastname and for apple sign in nothing gets saved even though when logging in I see the request for a name from apple sign in

Is there a way I can get the raw data that's being recieved from apple and google 0auth?

I'm defo deleting my user here: https://account.apple.com/account/manage and on pocketbase
And I see the email (which works great) and name being requested when I sign in.

thank you!

**Comment by Laestry** [comment]
I've figured it out by doing it with the manual code exchange, I could probably try to write an example and we could add it to the official docs? for both android and ios.

**Reply by leeoocca** [reply]
Can you post it here for now?

**Reply by Laestry** [reply, marked as answer]
> Can you post it here for now?

ahhaha sorry about that! classic meme "I've found the answer" and doesn't post the answer:
```
      final credential = await SignInWithApple.getAppleIDCredential(
        scopes: [
          AppleIDAuthorizationScopes.email,
          AppleIDAuthorizationScopes.fullName,
        ],
      );

      final authData = await pb.collection('users').authWithOAuth2Code(
        'oidc',
        credential.authorizationCode,
        '', // The code verifier
        redirectURL,
        createData: {"name": credential.givenName, "family": credential.familyName},
      );
```

it's pretty simple, I use https://pub.dev/packages/sign_in_with_apple but I don't remember why the code verifier needs to be empty

**Reply by ganigeorgiev** [reply]
Just for info in case someone stumble here, a small change was submitted in the `develop` branch in relation to https://github.com/pocketbase/pocketbase/issues/7090 that is supposed to "fix" and return the Apple user's name with the regular `authWithOAuth2` all-in-one call, aka:

```dart
final authData = await pb.collection('users').authWithOAuth2('apple', (url) async {
  await launchUrl(url);
});
```

Once again, I don't have Apple developer account and I haven't really verified it _(Apple remains the only provider that wasn't manually tested)_, so I'll appreciate if someone has free time and willing to test the `develop` branch.

Note also that while Apple is technically an OIDC provider, their implementation comes with [some differences](https://bitbucket.org/openid/connect/src/master/How-Sign-in-with-Apple-differs-from-OpenID-Connect.md) _(e.g. lack of userinfo endpoint, requirement about a POST redirect in case scopes are specified, etc.)_ and I'd recommend to use the dedicated `apple` provider instead.
